compiler/rustc_middle/src/ty/consts.rs

```diff
@@ -69,13 +71,25 @@ impl<'tcx> Const<'tcx> {
         }
     }
 
+    #[instrument(skip(tcx), level = "debug")]
     fn try_eval_lit_or_param(
         tcx: TyCtxt<'tcx>,
         ty: Ty<'tcx>,
         expr: &'tcx hir::Expr<'tcx>,
     ) -> Option<&'tcx Self> {
         let lit_input = match expr.kind {
             hir::ExprKind::Lit(ref lit) => Some(LitToConstInput { lit: &lit.node, ty, neg: false }),
+            hir::ExprKind::Block(
+                hir::Block {
+                    stmts: [],
+                    expr: Some(hir::Expr { kind: hir::ExprKind::Lit(ref lit), .. }),
+                    ..
+                },
+                _,
+            ) => {
+                // simplify blocks that consist of a single literal to a literal
+                Some(LitToConstInput { lit: &lit.node, ty, neg: false })
+            }
```

Comment:
this does not allow negated literals, it's probably best to move l114-l119 abovethe `lit_input` match:
```rust
// Unwrap a block, so that e.g. `{ P }` is recognised as a parameter. Const arguments
        // currently have to be wrapped in curly brackets, so it's necessary to special-case.
        let expr = match &expr.kind {
            hir::ExprKind::Block(block, _) if block.stmts.is_empty() && block.expr.is_some() => {
                block.expr.as_ref().unwrap()
            }
            _ => expr,
};
```
that would mean that we always optionally unwrap one brace, which seems sensible to me.

after that r=me